ow_power_system/src/power_system_node.cpp

```diff
@@ -129,23 +79,66 @@ int main(int argc, char* argv[]) {
   std_msgs::Float64 soc_msg;
   std_msgs::Float64 rul_msg;
   ROS_INFO ("Power system node running");
-
-  int power_lifetime = power_csv.size();
+  
   while (ros::ok()) {
-    if (t_step < power_lifetime) {
-      soc_msg.data=power_csv[t_step][voltage_i];
-      rul_msg.data=power_csv[t_step][rul_i];
-    } //publishes last voltage value if sim time exceeds the length of the csv
-
-    //publish current SOC
-    SOC_pub.publish(soc_msg);
-    RUL_pub.publish(rul_msg);
-    t_step+=1;
-    
-    ros::spinOnce();
-    rate.sleep();
-   
-  }
 
+    // For each line of data in the example file, run a single prediction step.
+    for (const auto & line : power_data) {
+      // Get a new prediction
+      auto prediction = prognoser->step(line);
+      // Get the event for battery EoD. The first line of data is used to initialize the observer,
+      // so the first prediction won't have any events.
+      if (prediction.getEvents().size() == 0) {
+          continue;
+      }
+      auto eod_event = prediction.getEvents().front();
+      // The time of event is a `UData` structure, which represents a data
+      // point while maintaining uncertainty. For the MonteCarlo predictor
+      // used by this example, the uncertainty is captured by storing the
+      // result of each particle used in the prediction.
+      UData eod_time = eod_event.getTOE();
+      if (eod_time.uncertainty() != UType::Samples) {
+        ROS_ERROR_NAMED("power_system_node", "Unexpected uncertainty type for EoD prediction");
+      }
+      
+      // For this example, we will print the median EoD.
+      auto samples = eod_time.getVec();
+      std::sort(samples.begin(), samples.end());
+      double eod_median = samples.at(samples.size() / 2);
+      auto now =  MessageClock::now();
+      auto now_s = duration_cast<std::chrono::seconds>(now.time_since_epoch());
+      double rul_median = eod_median - now_s.count();
+      rul_msg.data = rul_median;
```

Comment:
Last time I spoke with Chetan he mentioned that RUL was a calculation based partially on the SOC value. Does this new function take that into account? My PR also remaps the SOC topic and updates that with fault values when triggered, so that new fault value should be the new value used to RUL calculations (assuming that's still the case).

Reply:
From what I understand, RUL is based on the End of Discharge (EOD) calculation (simply EOD - current time), which is extracted from a separate state matrix than that which contains SOC. That being said, both are based on the power and voltage inputs, so they are certainly intertwined in ways that are hidden within the black box of GSAP. I had not discussed this with Chetan (mainly bc he was out most of the week), but I will touch base with him to clarify.